Add cq_cache decorator

Jojain opened the pull request:
Add a file based cache decorator function that allow to cache the result of a function creating a cadquery object.
When exploring design ideas from a complex geometry, rerunning the whole script at each change can be cumbersome. This plugin aims at helping on that.

It does it by writing a step file of the object created by decorated the function the first time a set of parameters is given and import this step if the script/the decorated function is run again with the same parameters.

While writing the tests I realized that the wrapper function doesn't necessarily return the same type of object as the decorated function. That's not goot.
I couldn't come up with a way to check the return object type of the decorated function so I'm asking help on this, if someone as any idea.

A solution could be to force the decorated function to return a specific type of object so the wrapper could return the same, but I think that makes the decorator less useful.

I'm happy to receive your feedback on this, thanks !

adam-urbanczyk commented:
You probably want to use internal data format of OCCT to have no losses. So `cq.Shape.exportBrep`, I'll add `cq.Shape.importBrep` to enable this.

Jojain commented:
Yes definitely, I searched for this on the cadquery doc but I'm not sure there info on it. I may just have missed it though.

jmwright commented:
[Docmentation link for exportBrep](https://cadquery.readthedocs.io/en/latest/classreference.html?highlight=exportBrep#cadquery.Shape.exportBrep)

Jojain commented:
Thanks I'll add this and I may have a solution for returning the right type of objects each time, will do that

Jojain commented:
This should be done now, I've switched to BREP files ( @adam-urbanczyk I wrote a importBrep function if you want to add it in cadquery) and I added some logic to be able to return the right type of object when decorating a function that returns a cadquery object.
I think I took care of everything but in the end it got a bit more complicated than expected so if you see any flaws let me know.

jmwright commented on plugins/cq_cache/README.md:
```diff
@@ -0,0 +1,67 @@
+# cq_cache
+
+This plugin provide a decorator function that allow you to add a file based cache to your functions to allow you to speed up the execution of your computational heavy cadquery functions.
```
```suggestion
This plugin provides a decorator function that allows you to add a file based cache to your functions to allow you to speed up the execution of your computationally heavy cadquery functions.
```

jmwright commented on plugins/cq_cache/README.md:
````diff
@@ -0,0 +1,67 @@
+To use this plugin after it has been installed, just import it and use the decorator on your functions
+
+```python
+#decorate your functions that build computational heavy shapes
````
```suggestion
# decorate your functions that build computationally heavy shapes
```

jmwright commented on plugins/cq_cache/README.md:
```diff
@@ -0,0 +1,67 @@
+    make_cube(1,1,1+i)
+
+clear_cq_cache()
+#>>> Cache cleared for 1.036 MB
```
```suggestion
# >>> Cache cleared for 1.036 MB
```

jmwright commented:
Thanks @Jojain this looks great. I think it would be helpful for there to be docstrings in all the `cq_cache.py` methods for us to reference in the future, and if I run black against the directory it still shows changes in `cq_cache.py`. I need to get the lint check fixed for PRs on this repo. I'll try to take a look at that tomorrow.

marcus7070 commented on plugins/cq_cache/cq_cache.py:
```diff
@@ -0,0 +1,138 @@
+        os.remove(os.path.join(CACHE_DIR_PATH, cache_file))
+    print(f"Cache cleared for {round(cache_size*1e-6,3)} MB ")
+
+def using_same_function(fct, file_name):
+    with open(file_name,"r") as f :
+        cached_function = "".join(f.readlines()[:-1])
+
+    caching_function = inspect.getsource(fct)
+    if cached_function == caching_function:
+        return True 
+    else :
+        return False
```
This code on stackexchange might be of interest to you: https://codereview.stackexchange.com/questions/192848/python-function-to-hash-a-python-function

The author does a similar task but he ignores comments, docstrings and minor formatting changes. Also stores the result as a small hash instead of the complete code.

It's a fair bit of extra work though, it's up to you if you want to implement it.

Jojain replied:
It doesn't seems too complicated so I might give it a try. Even though I'm not sure it will make more time gain overall than the one I'm going to take adding this 😄

marcus7070 replied:
Yeah, but I'll give you style points for implementing it. :rofl:

marcus7070 reviewed (commented):
I haven't had a chance to look at this in depth, but this is what I see from a quick look. Very interesting plugin @Jojain!

Jojain commented on plugins/cq_cache/cq_cache.py:
```diff
@@ -0,0 +1,138 @@
+    os.chdir(cwd)
+
+
+def build_file_name(fct, *args, **kwargs):
+    SPACER = "_"
+    file_name = fct.__name__
+    for arg in args:
+        file_name += SPACER + str(arg)
+    for kwarg_value in kwargs.values():
+        file_name += SPACER + str(kwarg_value)
+
+    return file_name
```
> basically you're just using memory location
I'm not sure what you mean here. However I agree that using hash is a good idea, I will look more in details how it can be used.

marcus7070 replied:
Sorry @Jojain, I read your code too quickly and I think I missed how you were using it. What types of arguments and keyword arguments do you intend to support in `build_file_name`?

I suspect you are going to say floats and bools, and explicitly not supporting complicated classes like Workplane or user defined classes. On my first look I saw the `CQ_TYPES = [cq.Shape, cq.Solid, cq.Shell, cq.Compound, cq.Face, cq.Wire, cq.Edge, cq.Vertex, TopoDS_Shape, cq.Workplane]` line at the top of your code and jumped to the conclusion that they were the types you were supporting, but I see that's not the case now.

My point is still the same though, you can't rely on `str(arg)` to uniquely identify an object. That's what `hash(arg)` is for (kind of).

I would recommend that this function firstly type check for any args/kwargs that are Workplanes and raise a `TypeError('Can not cache Workplane')`. Then hash every arg/kwargs, encode the hash with [base64.urlsafe_b64encode](https://docs.python.org/3/library/base64.html#base64.urlsafe_b64encode), and return that as the filename. 

The type check for Workplane is only required because `hash(cq.Workplane())` returns an int based off the id of the Workplane, which is useless for caching. Actually it's dangerous for file based caching, since between Python runs it's possible for a Workplane to wind up with the same hash/id (and string representation).

User defined classes without a `__hash__` method will run into the same problem as Workplane when used as args, but I'm not sure how to protect against that. Maybe just mention it in the doc string.

Ignore my question above @adam-urbanczyk, hashing Workplanes is not required.

Jojain commented:
@jmwright @marcus7070  I have added docstrings and tried to use hash instead of str for the identification of passed arguments.
I think the use of hashing for everything would probably better than the use of strings but honestly I don't really understand how it works and I don't want to spend time on trying to figure it out.
If someone more knowledgeable than me on this wants to implement more robust changes I see no problem.  
If you are good with the changes we can call it done.

(BTW I don't know what the lint test is about, I tried to run black and it still failing, is there a way to make it work ?)

jmwright commented:
@Jojain Thanks for all the work you've put into this. The black check here on GitHub always fails for PRs because of a missing feature in GitHub. One thing I have on my list is to run black manually in CI and see if I can get that to work instead of the current Action from the Marketplace.

jmwright commented:
@marcus7070 I'm ready to merge this. Let me know if you are too.

marcus7070 commented:
> @marcus7070 I'm ready to merge this. Let me know if you are too.

Not ready yet, I'm just exploring some of this hashing stuff. Python has deliberately broken the way I want to use the hash function, because repeatable hashes (when used in Python dicts) are a security vunerability. I'll try to push my changes later today.

marcus7070 reviewed (approved):
I pushed some changes, have a look @Jojain and let me know if you have any questions or changes. 

@jmwright after @Jojain has had a look (and yourself if you want) I'm happy for this to be merged.

Jojain commented on plugins/cq_cache/cq_cache.py:
```diff
@@ -74,24 +76,19 @@ def build_file_name(fct, *args, **kwargs):
+
+    # hash all relevant variables
+    hasher = hashlib.md5()
+    for val in [fct.__name__, args.__repr__(), kwargs.__repr__()]:
```
How is this different from my initial way of doing it using `str(arg)` ? 
From what I understand the difference with my initial code is that now instead of storing the `str(arg)` directly in the file name, you concatene everything is a resulting hash str. What for ?

marcus7070 replied:
> How is this different from my initial way of doing it using `str(arg)` ?

Yeah, not much difference. I had thought I could get it much more accurate, but the `hash` implementation is mainly for supporting dicts and isn't suitable. 

It's possible to hash the bytes output from pickle. This would be as accurate as I was hoping for, but you can't pickle most CQ types, so it isn't practical for us.

`repr` is slightly better than `str`, since:
```
>>> str(1) == str("1")
True
>>> repr(1) == repr("1")
False
```

I haven't solved the main problems with a `str(arg)` based caching though, mainly that any complicated object (including lambdas, generators, `cq.Plane` and `cq.Location`) will never compare the same and will be a seperate cache value each time the function is called. :shrug: 

> From what I understand the difference with my initial code is that now instead of storing the `str(arg)` directly in the file name, you concatene everything is a resulting hash str. What for ?

The hash string is a fixed length, this could easily blow out past filename limits with large args. Especially given it's going to be used on computationally heavy functions where you wouldn't be surprised to see a large array of (x, y, z) tuples as an argument.

Jojain replied:
Alright thanks for the explanation. It's a bit unfortunate that we can't support complicated types but maybe someone will have a better idea about this in the future !

marcus7070 replied:
I don't have much experience with pickle. I gave it a try and it errored on simple CQ objects. We might be able to get it working with [`__getstate__`](https://docs.python.org/3/library/pickle.html#object.__getstate__) methods on a few CQ classes.

Jojain commented:
I just had some questions but I'm good with the changes ! Thanks @marcus7070  you found a nice way to remove the use of eval. that's great !

Jojain commented:
We might also want to wait [https://github.com/CadQuery/cadquery/pull/735](url) to be resolved and use the importBrep implementation from cadquery ? (Altough it should be identical so we can merge it as it is)

marcus7070 commented:
> We might also want to wait [CadQuery/cadquery#735](https://github.com/CadQuery/cadquery/pull/735) to be resolved and use the importBrep implementation from cadquery ? (Altough it should be identical so we can merge it as it is)

I would say merge now and fix that later.

jmwright commented:
I see that @Jojain gave a thumbs-up on merging, so I'm going to go ahead and do that now. Thanks for the contribution @Jojain !
